classes/core/Core.php

```diff
@@ -498,18 +498,16 @@ private static function _getUrlComponents($urlInfo, $isPathInfo, $offset, $varNa
      */
     public static function classFromFile(SplFileInfo $file): string
     {
-        $pathFromBase = trim(Str::replaceFirst(base_path(), '', $file->getRealPath()), '/');
-        $libPath = 'lib/pkp/';
-        $namespace = Str::startsWith($pathFromBase, $libPath) ? 'PKP\\' : 'APP\\';
-
-        $path = $pathFromBase;
-        if ($namespace === 'PKP\\') {
-            $path = Str::replaceFirst($libPath, '', $path);
-        }
-        if (Str::startsWith($path, 'classes')) {
-            $path = Str::replaceFirst('classes/', '', $path);
+        $libPath = realpath(base_path(PKP_LIB_PATH));
+        $isLib = str_starts_with($file->getRealPath(), $libPath);
+        $className = str_replace($isLib ? $libPath : realpath(base_path()), '', $file->getRealPath());
+        $className = preg_replace('#^[\\\\/]classes|\.php$#', '', $className);
+        $className = str_replace('/', '\\', '/' . ($isLib ? 'PKP' : 'APP') . $className);
+
+        if (!class_exists($className)) {
+            error_log(new Exception("Failed to map the file \"{$file->getRealPath()}\" to a class"));
```

Comment:
Considering that we might still have files that are not namespaced, and that we're close to release the 3.4, I thought it's better to avoid surprises, so I just left the message with a helpful trace :)